03_08_02 Core 4.2 Discovery

marcelbuesing opened the pull request:
On top of the previous the PR (i recommend to merge #1 first) this adds device discovery as described in 03_08_02 Core 4.2 Discovery.
I have also added a example to show how it works. Tested this in my network and it worked fine. Even when setting the timeout down to 50 microseconds the MDT IP router still managed to answer in time. I think theoretically it could make sense to actually return a stream of SearchResponses and remove the timeout instead of the Vector, I assume that might make handling of this a little more cumbersome for people with no experience with Streams though.

The friendly name is encoded according to ISO_8859_1 and since this is not supported by the std lib I guess it could make sense to make the encoding crate a dependency to be able to properly encode/decode it. The CStr approach also worked returning a readable result but I assume for some characters that might not work.

marcelbuesing committed:
feat: Add device discovery

AndreaNiklaus commented:
Hi Marcel, thank you for your effort. Is this feature ready for the merge?
I think is possible to add another function to return results in a stream fashion in the future if needed.

marcelbuesing commented:
Almost, I will add one more small change for the ISO8850-1 part.

marcelbuesing commented:
Ok done now, made two more small changes.
I fixed a clippy lint that indicated an error in reading the ManufacturerDataDIB.
Besides that I also changed the friendly_name part, meaning the encoding crate is now a dependency. Not great as it is unmaintained, but I guess it does not make sense for users of the crate to have to deal with the encoding.
